text/differential-layer-format.md

```diff
@@ -0,0 +1,143 @@
+
+Delta layers contain a sorted, indexed list of all changes to pages in a
+certain segment and a range of LSNs, whereas Image layers contain a snapshot of
+the pages at a certain LSN.
+
+The Differential Layer format tries to improve and replace the Delta layer format.
+
+## Problems with Delta layer; and improvements
+
+### Page Version Branches
+As mentioned in "Storage and file formats; Structures", storing page versions
+in branches is an easy optimization which makes finding the right data easy.
+This does mean that we will have one full page image in potential overhead for
+each modified page in the layer, but that is not too bad (with Delta layers
+this is one FPW per page in the segment, per delta layer, or we'd have
+cross-file dependencies).
+
+As there may be multiple kinds of access patterns to pages, it is beneficial to
+store some metadata in the BranchPtr which discerns the kind of branch that
+pointer points to. This allows for compacter storage of the actual branches,
+at the cost of some extra data in the
+
+Not all branches will describe multiple page versions: It is indeed possible
+that a page will be updated only once during a checkpoint.
+
+In that case, it
+is quite wasteful to store the metadata of '0 page versions'
+
+- FPI, 0 revisions
+- will-init WAL record, 0 revisions
+- FPI-initiated branch, N WAL records depending on that FPI
+- WAL-record initiated branch, N WAL records depending on that WAL record
+- ...
+
+In the physical format, because the LSNs of the WAL records are sorted and stored
+in increasing order, we could apply compression to this array to reduce the
+storage overhead of the full branch structure.
+
+Additionally, because the length of the WAL record is stored in it's header, we
+don't actually need to store the length of the WAL record in the branch structure;
+we can read it directly from the header, saving 4 bytes per WAL record.
+
+With LSN array compression and the emission of the WAL record length also being
+stored outside the WAL record, we can reduce the storage overhead of non-FPI WAL
+records from 28 bytes per record down to less than 8 bytes per record, with an
+expected value of ~ 4 bytes (depending on how often a page is modified; more
+frequent / less noizy WAL record generation = lower overhead).
+
+
+
+### Page versions array
+The Delta layer uses an in-memory sorted array of <lsn, pageversion> pairs to
+represent the changes to a page. This is correct and allows for O(log N) version
+lookups (for N = number of total page versions). This is OK, but can be improved
+upon: as mentioned in "Storage and file formats; Structures", we always
+need a base page image to reconstruct a page. By limiting the array of page
+versions stored for each page in the Layer to Branches, we eliminate a
+significant amount of entries from the search space; improving this search speed
+by O(log fraction of non-image page-versions) and reducing memory usage.
+
```

Comment:
Right now delta layer stores only WAL data. Reconstructed pages are stored in image layer.
Delta layer may contains page images only if WAL contains FPI-s. Only in this case there can be multiple *branches* in inmem/delta layer. I do not understand what do you suggest to change in this schema.

Reply:
> WAL contains FPI-s

This is currently likely to happen, due to interplay of PG checkpointer, WAL-logging of hint bits, and the FPW being on by default.

> I do not understand what do you suggest to change in this schema.

What I suggest to change in this schema is to loosen the restriction to only allow WAL and WAL-FPIs in non-Image layers, so that we can store Branches instead, which are complete descriptors of the changes to a page over some period in time. This allows us to only store one pointer to that data in memory instead of the N pointers to the N changes to that page during that period.

Reply:
PG checkpoints can be actually disarmed (by setting max_wal_size to vey large value), and wal logging of hint bits IMHO also should be switched off.

So you suggest to include page images in delta layer (as part of *branch*)?
Having just one pointer to brnach in deta-layer header definitely makes sense.